src/Interpreters/executeQuery.cpp

```diff
@@ -1389,6 +1393,18 @@ std::pair<ASTPtr, BlockIO> executeQuery(
     ASTPtr ast;
     BlockIO res;
 
+    auto waitTime = ProfileEvents::global_counters[ProfileEvents::OSCPUWaitMicroseconds].load(std::memory_order_relaxed) - ProfileEvents::global_counters.getLastValue(ProfileEvents::OSCPUWaitMicroseconds);
+    auto busyTime = ProfileEvents::global_counters[ProfileEvents::OSCPUVirtualTimeMicroseconds].load(std::memory_order_relaxed) - ProfileEvents::global_counters.getLastValue(ProfileEvents::OSCPUVirtualTimeMicroseconds);
+    double min_ratio = context->getServerSettings().min_os_cpu_wait_time_ratio_to_throw;
+    double max_ratio = context->getServerSettings().max_os_cpu_wait_time_ratio_to_throw;
+    double current_ratio = (waitTime != 0 && busyTime == 0)
+        ? max_ratio
+        : std::min(std::max(min_ratio, static_cast<double>(waitTime) / static_cast<double>(busyTime)), max_ratio);
+    double probability_to_throw = (max_ratio == min_ratio) ? 0.0 : (current_ratio - min_ratio) / (max_ratio - min_ratio);
+
+    if (std::bernoulli_distribution server_overloaded(probability_to_throw); server_overloaded(thread_local_rng))
+        throw Exception(ErrorCodes::SERVER_OVERLOADED, "CPU is overloaded, latest wait time {}, latest busy time {}", waitTime, busyTime);
```

Comment:
We should try to make the exception message more user-readable. It should explain what is "wait time" and "busy time". Also, it can suggest reducing the query rate or increasing backoff on retries

Reply:
Done